_posts/2018-11-21-souper.md

```diff
@@ -0,0 +1,109 @@
+a chicken-and-egg problem.
+
+The solution to some manifestations of this problem is the use of
+superoptimizers. The academic discipline of superoptimizers is very
+old. The idea is to treat the code that was written as more
+of a restriction, a specification. The superoptimizer generates a ton
+of randomly generated native code and checks how it behaves
+differently than the written code. If it can generate a faster native
+code sequence than the compiler while keeping behavior exactly the
+same, it wins.
+
+This "exactly the same" part can be incredibly expensive if not done
+correctly. The computational effort involved has kept it from being
+used very often. Computers get faster and theorems get better, and we
+are now seeing superoptimization research begin again.
+
+One superoptimizer in particular, named [Souper](https://arxiv.org/abs/1711.04422), has
+reached maturity while interoperating with the industry standard
+code generator (LLVM) and the industry standard SMT engine (Z3). It has
+kickstarted a renewed faith in researchers that superoptimization is a reasonable
+policy. It can take the LLVM IR that a compiler was going to feed into LLVM, and
+computes better IR. This can sometimes take a lot of time, and the
+code emitted is the result of a process that isn't auditable. The pipeline is
+placing total faith in Souper for the correctness of generated
+code.
+
+It's mostly useful for compiler engineers to use to tell that
+optimizations were missed, and to identify how to fix that using
+conventional pattern matching. That said, Souper offers the ability to
+drop in for clang and to generate the code that is run. Some projects are
+eager to make any trade-offs for performance that are acceptable.
+Other projects may want to get a feel for how fast they could run
+if they were to invest making sure mono generates good code.
+If the compile time increase doesn't discourage them, many
+projects may find some benefit in such an optimizing compiler.
+
+I recommend that curious readers install Z3, get a checkout of
+
+[https://github.com/google/souper](https://github.com/google/souper),
+
+and complete the compilation process described in that documentation.
+
+When AOTing code with mono, they're going to want to pass the
+commandline flags named there into the `---aot=llvmopts=` argument.
+
+As of the time of this writing, that is
+
+```
+llvmopts="-load /path/to/libsouperPass.so -souper -z3-path=/usr/bin/z3" 
+``` 
+
+Mono will then allow Souper to step in during the middle of the LLVM
+compilation and try it's best at brute-forcing some better code. If
+there's anything short and fast that does the job better, it will be
+found.
+
+It is frankly amazing that Mono can get such extensive optimizations simply
+by compiling to LLVM IR. Without changing a single line of Mono's source, we
+changed our compilation pipeline in truly dramatic ways. This speaks to Mono's
+flexibility as an embedded runtime. Embedders using mono should consider
```

Comment:
> This speaks to Mono's flexibility as an embedded runtime.

Does it?  Feels more like it speak's to LLVM's flexibility as a compiler framework.